### PR TITLE
8292353: TableRow vs. TreeTableRow: inconsistent visuals in cell selection mode

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
The issue is caused by TreeTableRow incorrectly selected when cell selection mode is enabled.

Changes:
- modified TreeTableRow.updateSelection()

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8292353](https://bugs.openjdk.org/browse/JDK-8292353): TableRow vs. TreeTableRow: inconsistent visuals in cell selection mode

### Reviewers
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/875/head:pull/875` \
`$ git checkout pull/875`

Update a local copy of the PR: \
`$ git checkout pull/875` \
`$ git pull https://git.openjdk.org/jfx pull/875/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 875`

View PR using the GUI difftool: \
`$ git pr show -t 875`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/875.diff">https://git.openjdk.org/jfx/pull/875.diff</a>

</details>

### CODE REVIEW

File: modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableRow.java [hunk 1]
```diff
@@ -439,12 +439,16 @@ private void updateItem() {
...
 
-        boolean isSelected = getTreeTableView().getSelectionModel().isSelected(index);
-        if (isSelected() == isSelected) return;
+        TreeTableViewSelectionModel<T> sm = getTreeTableView().getSelectionModel();
+        if (sm == null) {
+            return;
+        }
 
-        updateSelected(isSelected);
+        boolean isSelected = !sm.isCellSelectionEnabled() && sm.isSelected(index, null);
```
Comment (kleopatra) [inline modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableRow.java:448]: this is different from TableRow (which uses isSelected(index)) - why?

And we don't seem to have complete test coverage for row selection (otherwise this issue would have shown somewhere :) - please add test/s that fail/pass before/after the fix, respectively

**Review by kleopatra — CHANGES_REQUESTED** [review]

File: modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableRow.java (hunk 1 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TreeTableRow.java]: updated and added tests.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeAndTableViewTest.java [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+import javafx.scene.control.TreeTableRow;
+import javafx.scene.control.TreeTableView;
+import javafx.scene.input.MouseEvent;
+import test.com.sun.javafx.scene.control.infrastructure.KeyModifier;
+import test.com.sun.javafx.scene.control.infrastructure.MouseEventFirer;
+import test.com.sun.javafx.scene.control.infrastructure.VirtualFlowTestUtils;
+
+/**
+ * Tests for:
+ * - cell seleciton logic JDK-8292353.
```
Comment (danielpeintner) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeAndTableViewTest.java:56]: Nit: seleciton -> selection

**Review by kleopatra — CHANGES_REQUESTED** [review]
Fix looks good (left a minor inline comment which you might follow or not :)

The tests, though, ... need some love:

As a general rule, it's preferred to assert a single state per test method instead of changing the state repeatedly inside the same method. We can see that something is wrong when having code comments like: 

    line 240:         assertFalse(row.isSelected()); // JDK-8292353 failure

    line 249:         assertFalse(row.isSelected()); // JDK-8292353 failure

running the test before the fix and the second failure is never reached. We _want_ to see the failures, all of them :)

Write a test as near to the cause as possible: here the cause is an incorrect selection state of the TreeTableRow under certain conditions. That's unrelated to any user interaction, so a setup faking such an interaction is a detour which might (or not) bring in their own trouble (aside: I do see the css parser warnings when running TreeAndTableViewTest inside Eclipse - that is [JDK-8291853](https://bugs.openjdk.org/browse/JDK-8291853) which we don't yet understand, except that something fishy might be going on ;) Best to use the most simple setup to test that exact state, here that would be an isolated Tree/TableRow thats configured with a Tree/TableView at some index and setup the view's selection state as needed, for examples see TreeTableRowTest or something like the following

    @Test
    public void testRowSelectionStateOnCellSelectionEnabledMultiple() {
        // configure view
        addColumns();
        tree.getSelectionModel().setCellSelectionEnabled(true);
        tree.getSelectionModel().setSelectionMode(SelectionMode.MULTIPLE);
        // configure tableRow        
        int selected = 1;
        cell.updateTreeTableView(tree);
        cell.updateIndex(selected);

        // select all columns if possible
        tree.getSelectionModel().select(selected, null);
        assertEquals("sanity: all columns selected", tree.getColumns().size(),
                tree.getSelectionModel().getSelectedCells().size());
        assertFalse("row must not be selected in cell selection mode", cell.isSelected());
    }

and one for the second failure before the fix plus (for coverage) the other combinations of cell selection enablement/selection mode :) These should reside in TreeTableRowTest, IMO, to be "near" the other row state tests.

For coverage, we should also have the corresponding tests for TableRow (also isolated, not wrapped into faked user interaction). They should reside in TableRowTest. Which we don't have yet, time to add it :)

My suggestions: 

- convert the monolithic test methods into separate tests, each without changing the state you want to test
- configure and test the selection state of an isolated row (without faking user interaction, similar to the example above)
- for TreeTableRow, add the tests to TreeTableRowTest
- for TableRow, add TableRowTest and add analogue methods as those you added for TreeTableRow
- remove TreeAndTableViewTest

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeAndTableViewTest.java (hunk 2 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeAndTableViewTest.java]: fixed, thanks!

**Comment by andy-goryachev-oracle** [comment]
Dear @kleopatra : thank you for insightful comments and suggestions!

I do agree that, in general, it might be beneficial to have one test case per condition or failure, but it should not be a requirement.  I believe it is perfectly fine to have a test that executes a complex scenario.  In this particular case, the failure is known, it is described in JDK-8292353, and the fact that there are multiple places where test fails (before the fix) is not important.  What is important is that the failing paths are exercised and pass as a result of the fix.

I do disagree with the requirement to write a highly artificial tests like testRowSelectionStateOnCellSelectionEnabledMultiple() because it makes too many assumptions on the internal structure of the code being tested.  I would rather have a test that simulates as much as possible the actual user interaction.  I am not saying we have to forbid simple tests, but rather we need both.  After all, we have all these tests and yet plenty of bugs avoided detection for years.  It took me 10 minutes to write and test a simple example and identify a number of issues (though many of them already in JBS) with Tree/TableView.

I'd say let's allow some diversity in our testing approaches.  Unless there is a *very* good reason to refactor, I'd rather leave TreeAndTableViewTest as is and not touch other files, as they are too big and I don't want to replicate code.

What do you think?

**Comment by kleopatra** [comment]
> 
> I do agree that, in general, it might be beneficial to have one test case per condition or failure, but it should not be a requirement. 

well, this project follows (should, at least, there are legacy areas ;) established best practices for unit testing. If you want that changed, the best place to discuss such a change in strategy is the mailinglist.

> I believe it is perfectly fine to have a test that executes a complex scenario. 

we certainly can have tests for complex scenarios - but only if testing such a complex scenario. That's not the case here.

> 
> I do disagree with the requirement to write a highly artificial tests like testRowSelectionStateOnCellSelectionEnabledMultiple() because it makes too many assumptions on the internal structure of the code being tested.

with all due respect: that sentence is completely wrong. 

- there is nothing artificial to test a class in the most minimal context possible, in fact that's the whole point of unit testing
- the amount of assumptions on internal structure is zero, they use public api to follow the three A (Arrange: create the cell, configure it at a fixed location inside a virtualized control) A (Act: change the selection of  the control) A (Assert: verify the cell has updated its own state accordingly)

> After all, we have all these tests and yet plenty of bugs avoided detection for years.

that's mostly because the coverage is .. suboptimal ;) As this issue demonstrates: there are no simple tests to guarantee a treeTableRow updates itself reliably.

>  I would rather have a test that simulates as much as possible the actual user interaction.

grabbing a tableRow from a temporary scenegraph (note: its scene is null once you get it), adding its tableCells to a different scenegraph (note: after mouseFirer constructor tableCell.getParent() != tableRow) and fire a mouseEvent onto that isolated tableCell is _near .. actual user interaction_? That's not case ;)

>  Unless there is a very good reason

The very good reason is that your test setup is suboptimal (apart from changing the recommended Attach-Act-Assert into Attach-Act-Assert-ActAgain-AssertAgain-... ) for the issue here ;)

We have a very clear functional path to test: a fully configured TableRow must sync its own selection state to the TableView's selection state at the location of the row. 

So all we need is a TableView (with columns, as we want cell selection) and a TableRow configured to represent a location in that TableView. Changing table's selection at the location of the row must change (or not, depending on selection context) the selection state of the row. The most simple means to change table's selection is programatically using public api on the table's selectionModel with a direct mapping of 

     selectionModel.select -> tableRow.updateSelection

With yours: 

     stageloaderA -> createAndConfigure tableRows/Cells  
                           -> tableRow = lookup row at index
                           -> stageloaderA.dispose
     // at this point we are already far from any"real" context - no user interaction possible
     tableCell = lookup column for tableColumn
     // nevertheless with rather normal micro-cosmos:
     assertEquals(tableCell.getParent(), tableCell.getTableRow());
     mouseFirer -> stageLoaderB -> insert targetNode into its own root
     // at this point we have an isolated tableCell in a slightly weird state, now failing the above assert:
     assertEquals(tableCell.getParent(), tableCell.getTableRow());
     // now fire a mouseEvent onto that isolated tableCell
     // (which nevertheless is fully configured with its tableView and location)    
     mouseFirer.doFirer -> tableCell.behaviour ->
                 __selectionModel.select -> tableRow.updateSelection__

and here at the very end we are again testing the same as in the simple case, after going on a detour which might (or not) have side-effects or bugs of their own. We might go that tour if we are specifically interested in that path, that is when we are really want to test the behaviour (mouse and keyboard input) - and then we must be certain that this very last expectation is fulfilled.

> 
> What do you think?

Same as yesterday:

- missing direct tests
- unrelated, overly complex (and potentially testing the wrong thingy) existing test

**Comment by andy-goryachev-oracle** [comment]
Dear @kleopatra : 

Thank you so much for the wisdom.  Let me try to refactor the test cases.

A few questions:

1. StageLoader - since it is being disposed of by MouseFirer, would it make sense to explain how to use it properly in its javadoc?  Perhaps a StageLoader needs to be created and explicitly disposed of at the beginning and end of each test?

2. I noticed that despite MouseFirer's firing MOUSE_PRESSED followed by MOUSE_RELEASED, the cell's pseudostyles contain "pressed" which seems to be incorrect.  Perhaps it is due to multiple stages being created?

3. Why do I need to call VirtualFlowTestUtils.getCell(table, 0) for the cells to be created?  Do you know what is missing here? 

thank you.

**Comment by andy-goryachev-oracle** [comment]
addressed feedback from @kleopatra : 

- implemented direct tests: TreeTableRowTest, TableRowTest
- properly using StageLoader
- tree table tests fail with the current master, pass with this PR.  also, tree/table behavior is ok with a standalone tester (not checked in)

Lessons learned:
- StageLoader is required in this case to properly initialize the cells.  No more calls to VirtualFlowTestUtils

I do want to keep my TreeAndTableViewTest as it exercises different paths, which is beneficial.  There will be more test cases applicable to both Tree- and TableView.

thank you.

**Comment by kevinrushforth** [comment]
I haven't looked at these specific tests in detail, but as a general practice I see value in having both types of tests: unit tests that test a specific thing, and tests of various scenarios.

**Comment by kleopatra** [comment]
Never thought that I would ever feel on the verge of arguing against adding tests ;) 

> I haven't looked at these specific tests in detail, but as a general practice I see value in having both types of tests: unit tests that test a specific thing, and tests of various scenarios.

Yeah, we could do additional testing of broader scenarios - assuming the tightly bounded specific tests are done. We should answer two questions:

- what _exactly_ do we expect to gain, that is which additional safety do they give?
- what is the boundary of those scenarios, why add one and not another? 

In this particular case we have a tighly bounded problem: the row updates itself incorrectly based on the state of the selection model. We know that all other potential collaborators talk to the selection model directly. Given we fixed the problem (done), and have failing/passing tests before/after the fix involving an isolated, fully configured cell (not yet done), my answers:

- nothing I can see (might be missing something, of course :)
- the choice of scenarios (scene only, simulating mouse input) feels arbitrary, why not key input (or touch, accessiblity .. )? They all are at the same level of unrelatedness, IMO - we are not testing the effectiveness of this bug fix, but something else (for the mouse input f.i. tableCell - behaviour)

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/ControlUtils.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    public static TableRow getTableRow(TableView t, int row) {
+        for (Node n: t.lookupAll(".table-row-cell")) {
+            if (n instanceof TableRow c) {
+                if (row == c.getIndex()) {
+                    return c;
+                }
+            }
+        }
+        throw new Error("TableRow not found at " + row);
+    }
```
Comment (kleopatra) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/ControlUtils.java:157]: no review, just a couple of comments:

Generally, should prefer api over lookup - we have shims to access package/protected api. 

In this particular case, the lookup is brittle: it assumes that

- either there is a single row configured with the given index 
- or that it's the first when iterating over the set

With the first being incorrect (corner case, of course :)

    @Test
    public void testLookupLastNotEmptyCell() {
        stageLoader = new StageLoader(table);
        int index = table.getItems().size()- 1;
        Set<Node> tableRows = table.lookupAll(".table-row-cell").stream()
                .filter(n -> n instanceof TableRow<?> && ((TableRow<?>) n).getIndex() == index)
                .collect(Collectors.toSet());
        assertEquals(1, tableRows.size());
    }

we fall back to the second assumption, which is implementation dependent (and unspecified) - it's accidental that we actually do get the row we want.

Having done a bit of digging into VirtualFlowTestUtils (see [JDK-8293356](https://bugs.openjdk.org/browse/JDK-8293356)) I now think it's safe enough to use it (even though it's on the oldish side, not using recently added api, probably needs some polish) - as long as  we keep the scenegraph stable during the test, f.i. by manually adding the table to a stage (either via StageLoader or showing in our own).

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/ControlUtils.java:157]: Is it possible to have two or more rows with the same row index?  i would imagine that will break a lot of things.

I am not sure why you think the lookup is brittle - after all, VirtualFlowTestUtils uses lookup to get a pointer to the VirtualFlow, is it not (line 333)?  Is it because the lookup may not return a cell if it is outside of the current view port area and therefore has not been created (while the .getCell() guarantees to create one)? 

Is this the only objection to the changes in this PR?  I suppose i could change the tests to use VirtualFlowTestUtils, though my preference would be still to test close to the real world scenario, using public APIs (as opposed to getting into internals using some internal test utils), given the fact that I *do* expect these cells to be visible given the dimensions of the table and columns.

Comment (kleopatra) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/ControlUtils.java:157]: > Is it possible to have two or more rows with the same row index? i would imagine that will break a lot of things.

don't imagine, verify ;) What happens when you run the test? What happens when you look at the scenegraph (f.i. with ScenicView)? Don't know why they are there but they are - well outside the viewport.

> 
> Is this the only objection to the changes in this PR?

I honestly cannot understand how you possibly could have reached the conclusion that this is the _only_ objection given my comments .. 

Anyway, I'm off here - don't know how to abort a review, but not going to waste more time and nerves on who doesn't seem to listen (that's my personal feeling, of course ;)

Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/ControlUtils.java:157]: sorry, what I really meant is - are the code changes ok and we are only discussing updating tests?

my goal here is to fix as many bugs as I possibly can, while listening to the feedback of more experienced team members and addressing their concerns.  it is not always clear when an exchange of ideas ends and marching orders begin, especially in situations where several possibilities exist.  perhaps a clearly annunciated request for specific changes would work?

I feel really uncomfortable with thousands of defects logged against openjfx, and all I want is to help.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java [hunk 4]
```diff
@@ -820,4 +834,71 @@ public class TreeTableRowTest {
...
+
+        assertTrue(c0.isSelected());
+        assertTrue(c1.isSelected());
+        assertTrue(c2.isSelected());
+        assertFalse(row.isSelected()); // JDK-8292353 failure
+    }
+
+    /** TreeTableView with cell selection enabled should not select TreeTableRows */
+    @Test
+    public void test_TreeTableView_jdk_8292353_select_all_but_one() {
```
Comment (aghaisas) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java:875]: This test passes with or without the code changes.
If test intention is to increase coverage, then the comment `// JDK-8292353 failure` needs to be removed. If not, the test needs some correction.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java [hunk 5]
```diff
@@ -820,4 +834,71 @@ public class TreeTableRowTest {
         TreeTableRow cell = new TreeTableRow();
         cell.setSkin(new TreeTableRowSkin(cell));
     }
+
+    /** TreeTableView with cell selection enabled should not select TreeTableRows */
+    @Test
+    public void test_TreeTableView_jdk_8292353_select_all() {
```
Comment (aghaisas) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java:840]: This test passes with or without the code changes.
If test intention is to increase coverage, then the comment `// JDK-8292353 failure` needs to be removed. If not, the test needs some correction.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeAndTableViewTest.java [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+    StageLoader stageLoader;
+
+    @After
+    public void after() {
+        if (stageLoader != null) {
+            stageLoader.dispose();
+        }
+    }
+
+    /** TreeTableView with cell selection enabled should not select TreeTableRows */
```
Comment (aghaisas) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeAndTableViewTest.java:61]: Comment needs correction. `TreeTableView` should be `TableView` & `TreeTableRows` should be `TableRows`

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/ControlUtils.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+        MouseEventFirer m = new MouseEventFirer(target);
+        m.fireMousePressAndRelease(modifiers);
+        m.dispose();
+
+        Toolkit.getToolkit().firePulse();
+    }
+
+    /**
+     * Finds a Node given the selector and predicate filter, then insures there is
+     * one one such node
```
Comment (aghaisas) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/ControlUtils.java:201]: Comment has a typo : "one one"?

**Review by aghaisas — CHANGES_REQUESTED** [review]

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java (hunk 4 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java:875]: I don't understand.  It fails for me

```
TreeTableRowTest > test_TreeTableView_jdk_8292353_select_all_but_one FAILED
    org.opentest4j.AssertionFailedError: expected: <false> but was: <true>
        at app//org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
        at app//org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:40)
        at app//org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:35)
        at app//org.junit.jupiter.api.Assertions.assertFalse(Assertions.java:227)
        at app//test.javafx.scene.control.TreeTableRowTest.test_TreeTableView_jdk_8292353_select_all_but_one(TreeTableRowTest.java:902)
```

command line:

```
/bin/sh gradlew -PSTUB_RUNTIME=../caches/modular-sdk -PFULL_TEST=true -PUSE_ROBOT=true --no-daemon --info --stacktrace cleanTest :controls:test --tests TreeTableRowTest -x :web:test --stacktrace 2>&1 | tee ~/`date +"test-%Y-%m%d-%H%M%S"`.log
```

test and ControlUtils supplanted on top of the current master.

did you switch and rebuild?

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java (hunk 5 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java:840]: similarly, this test also fails without code changes:

```
TreeTableRowTest > test_TreeTableView_jdk_8292353_select_all FAILED
    org.opentest4j.AssertionFailedError: expected: <false> but was: <true>
        at app//org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
        at app//org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:40)
        at app//org.junit.jupiter.api.AssertFalse.assertFalse(AssertFalse.java:35)
        at app//org.junit.jupiter.api.Assertions.assertFalse(Assertions.java:227)
        at app//test.javafx.scene.control.TreeTableRowTest.test_TreeTableView_jdk_8292353_select_all(TreeTableRowTest.java:870)
```

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeAndTableViewTest.java (hunk 6 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeAndTableViewTest.java:61]: fixed, thank you.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/ControlUtils.java (hunk 7 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/ControlUtils.java:201]: only one.  fixed.  thank you.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java (hunk 4 above)
Comment (aghaisas) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java:875]: A clean build and run behaves correctly. This test fails without code changes and passes with them.
Sorry for the confusion!

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java (hunk 5 above)
Comment (aghaisas) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java:840]: A clean build and run behaves correctly. This test fails without code changes and passes with them.
Sorry for the confusion!

**Review by aghaisas — APPROVED** [review]

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java (hunk 4 above)
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableRowTest.java:875]: Thank you, Ajit!

**Comment by kleopatra** [comment]
> /reviewer remove kleopatra

just noticed that this removal request didn't work (a bit unexpected ;) - @kevinrushforth please remove me as reviewer

**Comment by kevinrushforth** [comment]
> > /reviewer remove kleopatra
> 
> just noticed that this self removal request didn't work (a bit unexpected ;) - @kevinrushforth please remove me as reviewer for this PR

There's nothing to remove, at least from a Skara point of view. You aren't listed as a `Reviewer` of this PR.

If you mean that you are tagged as a _potential_ reviewer from GitHub's point of view, yes, since you left review comments. It doesn't translate into your having reviewed it, though. I can try to remove you as a potential reviewer, but that's under control of GitHub not Skara.

**Review by kevinrushforth — COMMENTED** [review]
The fix looks good.

As for the tests, I somewhat agree with the comments left by @kleopatra, especially the comment about the suitability of the tests in `TreeAndTableViewTest`. Having said that, they aren't harmful, so I will leave it up to you as to whether to remove that test class or leave it in.

I left a few specific comments on the tests inline. You can either update this PR (probably best) or file a follow-up test bug, which can be done during an upcoming test sprint.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewRowTest.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        // select 0:0 and 0:2
+        sm.select(0, null);
+        sm.clearSelection(0, col1);
+
+        assertTrue(c0.isSelected());
+        assertFalse(c1.isSelected());
+        assertTrue(c2.isSelected());
+        assertFalse(row.isSelected());
+    }
+}
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewRowTest.java:120]: yes, these cases are covered in TableViewSelectionModelImplTest

**Comment by andy-goryachev-oracle** [comment]
I give up - removing TreeAndTableViewTest.

Thank you @kevinrushforth for insights and suggestions!

**Review by kevinrushforth — APPROVED** [review]
Looks good.

**Review by aghaisas — APPROVED** [review]
